be/src/exec/pipeline/scan/morsel.cpp

```diff
@@ -111,6 +144,65 @@ StatusOr<MorselPtr> FixedMorselQueue::try_get() {
     }
 }
 
+PartitionSequenceMorselQueue::PartitionSequenceMorselQueue(MorselQueuePtr&& morsel_queue)
+        : _morsel_queue(std::move(morsel_queue)) {}
+
+std::vector<TInternalScanRange*> PartitionSequenceMorselQueue::olap_scan_ranges() const {
+    return _morsel_queue->olap_scan_ranges();
+}
+
+bool PartitionSequenceMorselQueue::empty() const {
+    return _unget_morsel == nullptr && _morsel_queue->empty();
+}
+
+StatusOr<MorselPtr> PartitionSequenceMorselQueue::try_get() {
+    if (_unget_morsel != nullptr) {
+        return std::move(_unget_morsel);
+    }
+    if (_morsel_queue->empty()) {
+        return nullptr;
+    }
+    ASSIGN_OR_RETURN(auto morsel, _morsel_queue->try_get());
+    auto* m = down_cast<ScanMorsel*>(morsel.get());
+    auto owner_id = m->owner_id();
+    ASSIGN_OR_RETURN(int64_t next_owner_id, _next_sequence_id());
+    _ticket_checker->enter(owner_id, next_owner_id != owner_id);
+    _current_sequence = owner_id;
+    return morsel;
+}
+
+std::string PartitionSequenceMorselQueue::name() const {
+    return fmt::format("partition_morsel_queue({})", _morsel_queue->name());
+}
+
+StatusOr<bool> PartitionSequenceMorselQueue::ready_for_next() const {
+    if (_current_sequence < 0) {
+        return true;
+    }
+
+    ASSIGN_OR_RETURN(int64_t next_sequence_id, _next_sequence_id());
+    if (next_sequence_id == _current_sequence) {
+        return true;
+    }
+
+    if (_ticket_checker->are_all_left(_current_sequence)) {
+        return true;
+    }
+
+    return false;
+}
+
+StatusOr<int64_t> PartitionSequenceMorselQueue::_next_sequence_id() const {
```

Comment:
it would be better to peek_sequence_id instead of next_sequence_id;
use StatusOr<bool>